### ISSUE TITLE
is it possible to have <j/> boundaries without spaces?

### ISSUE THREAD

**Issue by jonorthwash** [apertium/apertium-separable#11]
With the following rule, I'm trying to get "year-old", but instead get "year - old" (with spaces).

Rule:
```xml
<e lm="year-old" c="жашар">
	<p>
		<l>year<s n="n"/><s n="sg"/><j/>-<s n="guio"/><j/>old<s n="adj"/><s n="sint"/></l>
		<r>year-old<s n="adj"/></r>
	</p>
	<i><j/></i>
</e>
```

Example input and output:

```
Азамат алты жашар кичинекей бала.
```
↓  tagging, transfer
```
^Azamat<np><ant><m><sg>$ ^be<vbser><pres><p3><sg>$ ^the<det><def><sp>$ ^six<num><pl>$ ^year-old<adj>$ ^little<adj>$ ^kid<n><sg>$^.<sent>$^.<sent>$
```
↓ revautoseq
```
^Azamat<np><ant><m><sg>$ ^be<vbser><pres><p3><sg>$ ^the<det><def><sp>$ ^six<num><pl>$ ^year<n><sg>$ ^-<guio>$ ^old<adj><sint>$ ^little<adj>$ ^kid<n><sg>$^.<sent>$^.<sent>$
```
↓ generation
```
Azamat is the six year - old little kid.
```

**Comment by mr-martian** [comment]
The code causing this behavior appears to be

https://github.com/apertium/lttoolbox/blob/496aa8ebf9f9691f8027f957537c2b79870c5eed/lttoolbox/fst_processor.cc#L1044-L1053

So this would require having the transducer somehow encode that certan `$`s should not have spaces after them.

**Comment by mr-martian** [comment]
Now that I've redone the parser, this should actually be fairly straightforward. Word boundaries where you don't want a blank immediately afterwards can have `<$><->` instead of just `<$>`.

How do you want to write it? `<j space="no"/>` ?

**Comment by ftyers** [comment]
No!

**Comment by mr-martian** [comment]
No as in that's a terrible transducer hack or no as in that's a terrible xml representation?

**Comment by ftyers** [comment]
Here are the remaining letters: `c d f h k n o q u v x y z`

**Comment by mr-martian** [comment]
Except that right now `<j/>` is "boundary with space" and what is wanted here is "boundary without space".

There's only about 15 `.lsx` files in use, so it wouldn't be impossible to replace `<j/>` with something else and have `<j/>` mean `+` as elsewhere.

**Comment by ftyers** [comment]
Yes, that's what I think we should do. `<j/>` should mean boundary without space, and then we need another letter for boundary with space. It won't be a big deal to search and replace on the existing files.

**Comment by mr-martian** [comment]
Ok, but the specific issue here is a distinction between end of word with and without blank.

`<d/>` for "delimiter" perhaps?

Though that still leaves the question of how do we write a word boundary without a blank. It's not nearly as common so `<ds/>` could be reasonable, or an attribute as I suggested above, or we could eat up 2 of those letters, though that seems excessive.

**Comment by ftyers** [comment]
without a blank: `<j/>` with a blank `<d/>` sounds ok!

**Comment by mr-martian** [comment]
Except that to be consistent `<j/>` should be `+` as in monodix, rather than `<$>` as it currently is in lsx.

Here's all the symbols under discussion:

| thing | surface | transducer | current xml | proposed xml |
| -- | -- | -- | -- | -- |
| space | ` ` | ` ` | `<b/>` | `<b/>` (as everywhere) |
| compound mark | `+` | `+` | `+` | `<j/>` (as in monodix) |
| word boundary followed by a blank | `$ ^` | `<$>` | `<j/>` | `<d/>` |
| word boundary not followed by a blank | `$^` | `<$><->` | (none) | `<ds/>` maybe? or `<d space="no"/>`|

**Comment by ftyers** [comment]
How about `<j/>` is `+` inside `<mlu>` but `$^` outside it. I don't like the `<$><->` thing there shouldn't be a need for two symbols, just having `<->` would suffice.

**Comment by mr-martian** [comment]
Given that LUs can cross `<par>` boundaries, writing them inside `<mlu> ... </mlu>` seems rather complicated.

Regarding the transducer encoding, having it as one symbol would definitely be doable, I just have to write the parsing loop a bit differently. Or it could be `<$>` vs `<$->`.
